server: TestServerController failed

server.TestServerController [failed](https://mesolite.cluster.engflow.com/invocations/default/ddfb3951-050d-4e54-b8c0-67720b811feb?testReportRun=1&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cvc2VydmVyOnNlcnZlcl90ZXN0) on release-25.2 @ [0bce347cfb34e4cb99390673b0323d81ada697b3](https://github.com/cockroachdb/cockroach/commits/0bce347cfb34e4cb99390673b0323d81ada697b3):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2aad7ae]

```
Stack: 

```
goroutine 10586 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc0014a2e90?, {0x7327030, 0xc002f00c80})
	pkg/util/stop/stopper.go:227 +0x65
panic({0x5bd61c0?, 0xaf1ff80?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/cockroach/pkg/sql/stats.(*TableStatisticsCache).Stop(0xc005abd888)
	pkg/sql/stats/stats_cache.go:307 +0x6e
github.com/cockroachdb/cockroach/pkg/sql/stats.NewTableStatisticsCache.func1()
	pkg/sql/stats/stats_cache.go:271 +0x17
github.com/cockroachdb/cockroach/pkg/util/stop.CloserFn.Close(0x10?)
	pkg/util/stop/stopper.go:104 +0x12
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).AddCloser(0xc00818f4d0, {0x72ce620, 0xc0061a69b0})
	pkg/util/stop/stopper.go:265 +0x92
github.com/cockroachdb/cockroach/pkg/sql/stats.NewTableStatisticsCache({0x7329858, 0xc007483360}, 0xc005380000, {0x73267b0, 0xc0081c1b00}, 0xc00818f4d0, 0xc0037f1340)
	pkg/sql/stats/stats_cache.go:270 +0x198
github.com/cockroachdb/cockroach/pkg/server.newSQLServer({_, _}, {{{{{...}, 0x0}}, {{{...}, 0x0}}, {{{...}, 0x0}}, 0xc003ad7a40, 0x6999700, ...}, ...})
	pkg/server/server_sql.go:1041 +0x3b58
github.com/cockroachdb/cockroach/pkg/server.newTenantServer({_, _}, _, {0xc003c50280, 0xc005380000, 0xc002810c60, 0xc0056f1c20, 0xc002fffbc0, 0xc0046f0120, {0xc002810c60, ...}, ...}, ...)
	pkg/server/tenant.go:424 +0x9e8
github.com/cockroachdb/cockroach/pkg/server.newSharedProcessTenantServer({_, _}, _, {0xc003c50280, 0xc005380000, 0xc002810c60, 0xc0056f1c20, 0xc002fffbc0, 0xc0046f0120, {0xc002810c60, ...}, ...}, ...)
	pkg/server/tenant.go:254 +0x137
github.com/cockroachdb/cockroach/pkg/server.newTenantServerInternal({_, _}, {0xc003c50280, 0xc005380000, 0xc002810c60, 0xc0056f1c20, 0xc002fffbc0, 0xc0046f0120, {0xc002810c60, {0x72ce640, ...}, ...}, ...}, ...)
	pkg/server/server_controller_new_server.go:153 +0x158
github.com/cockroachdb/cockroach/pkg/server.(*topLevelServer).newTenantServer(_, {_, _}, _, _, _, {{0x0, 0x0}, {0x0}, 0x0, ...})
	pkg/server/server_controller_new_server.go:86 +0x398
github.com/cockroachdb/cockroach/pkg/server.(*serverController).newServerForOrchestrator(0xc004b8e1c0, {0x7327030, 0xc0056f0a50}, 0xc0014a2f10, 0xc00818f4d0)
	pkg/server/server_controller.go:407 +0x17e
github.com/cockroachdb/cockroach/pkg/server.(*channelOrchestrator).startControlledServer.func5.2(0xc001b74ee8, 0xc0014a2ff0, 0xc0056f0a50?, 0xc0061e1bc8, {0x7329858, 0xc00712d0e0}, 0xc00818f290, 0xc003a49ab0)
	pkg/server/server_controller_channel_orchestrator.go:369 +0x7b
github.com/cockroachdb/cockroach/pkg/server.(*channelOrchestrator).startControlledServer.func5({0x0?, 0x7327030?})
	pkg/server/server_controller_channel_orchestrator.go:393 +0x68e
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2({0x65fdaf6, 0x15})
	pkg/util/stop/stopper.go:507 +0x1e2
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 10485
	pkg/util/stop/stopper.go:496 +0x466
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestServerController
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestServerController2028214671
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestServerController.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-63373

server.TestServerController [failed](https://mesolite.cluster.engflow.com/invocations/default/b3b0e370-77ed-4c18-a018-793b5211a37c?testReportRun=4&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cvc2VydmVyOnNlcnZlcl90ZXN0) on release-25.2 @ [7e511ccc433217c8a182f3a7ece926a360a7cfa6](https://github.com/cockroachdb/cockroach/commits/7e511ccc433217c8a182f3a7ece926a360a7cfa6):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2aad7ae]

```
Stack: 

```
goroutine 10740 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc00186ccb0?, {0x7327030, 0xc00176a5f0})
	pkg/util/stop/stopper.go:227 +0x65
panic({0x5bd61c0?, 0xaf1ff80?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/cockroach/pkg/sql/stats.(*TableStatisticsCache).Stop(0xc004e08e08)
	pkg/sql/stats/stats_cache.go:307 +0x6e
github.com/cockroachdb/cockroach/pkg/sql/stats.NewTableStatisticsCache.func1()
	pkg/sql/stats/stats_cache.go:271 +0x17
github.com/cockroachdb/cockroach/pkg/util/stop.CloserFn.Close(0x10?)
	pkg/util/stop/stopper.go:104 +0x12
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).AddCloser(0xc003d55830, {0x72ce620, 0xc001a22ea0})
	pkg/util/stop/stopper.go:265 +0x92
github.com/cockroachdb/cockroach/pkg/sql/stats.NewTableStatisticsCache({0x7329858, 0xc004ba94a0}, 0xc004dea000, {0x73267b0, 0xc0014af5f0}, 0xc003d55830, 0xc004fffad0)
	pkg/sql/stats/stats_cache.go:270 +0x198
github.com/cockroachdb/cockroach/pkg/server.newSQLServer({_, _}, {{{{{...}, 0x0}}, {{{...}, 0x0}}, {{{...}, 0x0}}, 0xc0034b41e0, 0x6999700, ...}, ...})
	pkg/server/server_sql.go:1041 +0x3b58
github.com/cockroachdb/cockroach/pkg/server.newTenantServer({_, _}, _, {0xc005c86640, 0xc004dea000, 0xc0076f9320, 0xc004212780, 0xc004da7ba0, 0xc00512ad20, {0xc0076f9320, ...}, ...}, ...)
	pkg/server/tenant.go:424 +0x9e8
github.com/cockroachdb/cockroach/pkg/server.newSharedProcessTenantServer({_, _}, _, {0xc005c86640, 0xc004dea000, 0xc0076f9320, 0xc004212780, 0xc004da7ba0, 0xc00512ad20, {0xc0076f9320, ...}, ...}, ...)
	pkg/server/tenant.go:254 +0x137
github.com/cockroachdb/cockroach/pkg/server.newTenantServerInternal({_, _}, {0xc005c86640, 0xc004dea000, 0xc0076f9320, 0xc004212780, 0xc004da7ba0, 0xc00512ad20, {0xc0076f9320, {0x72ce640, ...}, ...}, ...}, ...)
	pkg/server/server_controller_new_server.go:153 +0x158
github.com/cockroachdb/cockroach/pkg/server.(*topLevelServer).newTenantServer(_, {_, _}, _, _, _, {{0x0, 0x0}, {0x0}, 0x0, ...})
	pkg/server/server_controller_new_server.go:86 +0x398
github.com/cockroachdb/cockroach/pkg/server.(*serverController).newServerForOrchestrator(0xc0076a4fc0, {0x7327030, 0xc0084633b0}, 0xc00186ccf0, 0xc003d55830)
	pkg/server/server_controller.go:407 +0x17e
github.com/cockroachdb/cockroach/pkg/server.(*channelOrchestrator).startControlledServer.func5.2(0xc0076a15d8, 0xc00186cd70, 0xc0084633b0?, 0xc005255bc8, {0x7329858, 0xc00515f5e0}, 0xc003d55560, 0xc005ba1e30)
	pkg/server/server_controller_channel_orchestrator.go:369 +0x7b
github.com/cockroachdb/cockroach/pkg/server.(*channelOrchestrator).startControlledServer.func5({0x0?, 0x7327030?})
	pkg/server/server_controller_channel_orchestrator.go:393 +0x68e
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2({0x65fdaf6, 0x15})
	pkg/util/stop/stopper.go:507 +0x1e2
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 10642
	pkg/util/stop/stopper.go:496 +0x466
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestServerController
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestServerController1752044911
    test_log_scope.go:76: use -show-logs to present logs inline
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=4</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestServerController.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/stats/stats_cache.go
+++ b/pkg/sql/stats/stats_cache.go
@@ -336,9 +336,16 @@ func (sc *TableStatisticsCache) Clear() {
 func (sc *TableStatisticsCache) Stop() {
 	sc.mu.Lock()
 	defer sc.mu.Unlock()
-	// Clear the cache first so that OnEvictedEntry calls shrinkEntryLocked
-	// for each entry, properly releasing the bytes tracked by the account.
-	sc.mu.cache.Clear()
+	// The stopper closer that calls Stop is registered in
+	// NewTableStatisticsCache *before* sc.mu.cache is initialised. If the
+	// stopper begins draining closers during cache construction (rare, but
+	// observed in TestServerController on shared-process tenants), Stop can
+	// fire on a partially-initialised cache. Guard against that here.
+	if sc.mu.cache != nil {
+		// Clear the cache first so that OnEvictedEntry calls shrinkEntryLocked
+		// for each entry, properly releasing the bytes tracked by the account.
+		sc.mu.cache.Clear()
+	}
 	sc.mu.acc.Clear(context.Background())
 	if sc.mon != nil {
 		sc.mon.Stop(context.Background())

